### PR TITLE
Enhanced Scatter Search

### PR DESCRIPTION
**Pull request by dweindl**
Adds an Enhanced Scatter Search optimizer.

So far, non-cooperative, non-self-adaptive. To be extended later.

### CODE REVIEW

File: pypesto/optimize/ess.py [hunk 1]
```diff
@@ -0,0 +1,662 @@
...
+            return False
+
+        if self.evaluator.n_eval >= self.max_eval:
+            self.exit_flag = ESSExitFlag.MAX_EVAL
+            return False
+
+        return True
+
+    def _combine_solutions(self) -> Tuple[np.array, np.array]:
+        """Combine solutions and evaluate."""
```
Comment (yannikschaelte) [inline pypesto/optimize/ess.py:484]: some more comments on the algorithm would be good.

File: pypesto/optimize/ess.py [hunk 2]
```diff
@@ -0,0 +1,662 @@
...
+    def _maybe_update_global_best(self, x, fx):
+        if fx < self.fx_best:
+            self.x_best = x[:]
+            self.fx_best = fx
+            self.x_best_has_changed = True
+
+    def _go_beyond(self, x_best_children, fx_best_children):
+        """Apply go-beyond strategy.
+
+        See [Egea2009]_ algorithm 1
```
Comment (yannikschaelte) [inline pypesto/optimize/ess.py:597]: In this algorithms, beyond "See xyz", always a short motivation+summary of why+what is happening would be good.

File: pypesto/optimize/ess.py [hunk 3]
```diff
@@ -0,0 +1,662 @@
...
+                    high=x_child,
+                )
+                x_parent = x_child
+                fx_parent = fx_child
+                x_child = x_new
+                fx_child = self.evaluator.single(x_child)
+
+                improvement += 1
+                if improvement == 2:
+                    Lambda /= 2
```
Comment (yannikschaelte) [inline pypesto/optimize/ess.py:627]:
```suggestion
            lambda_ = 1
            while fx_child < fx_parent:
                # update best child
                x_best_children[i] = x_child
                fx_best_children[i] = fx_child

                # create new solution, child becomes parent
                x_new = np.random.uniform(
                    low=x_child - (x_parent - x_child) / lambda_,
                    high=x_child,
                )
                x_parent = x_child
                fx_parent = fx_child
                x_child = x_new
                fx_child = self.evaluator.single(x_child)

                improvement += 1
                if improvement == 2:
                    lambda_ /= 2
```

File: pypesto/optimize/ess.py [hunk 4]
```diff
@@ -0,0 +1,662 @@
...
+        # Number of function evaluations at which the last local search took
+        #  place
+        self.last_local_search_neval = 0
+        # Whether self.x_best has changed in the current iteration
+        self.x_best_has_changed = False
+        self.exit_flag = ESSExitFlag.DID_NOT_RUN
+        self.evaluator = None
+        self.starttime = None
+
+    def minimize(self, problem: Problem, startpoint_method: StartpointMethod):
```
Comment (yannikschaelte) [inline pypesto/optimize/ess.py:369]:
```suggestion
    def minimize(self, problem: Problem, startpoint_method: StartpointMethod) -> pypesto.Result:
```

**Review by yannikschaelte — CHANGES_REQUESTED** [review]

File: pypesto/optimize/ess.py [hunk 5]
```diff
@@ -0,0 +1,662 @@
...
+
+import numpy as np
+
+import pypesto.optimize
+from pypesto import OptimizerResult, Problem
+from pypesto.startpoint import StartpointMethod
+
+logger = logging.getLogger(__name__)
+
+__all__ = ['ESSOptimizer', 'ESSExitFlag']
```
[the remark this answers is not in the document]
Comment (dweindl) [reply pypesto/optimize/ess.py:43]: Can do

File: pypesto/optimize/ess.py [hunk 6]
```diff
@@ -0,0 +1,662 @@
...
+        self.reset_round_counter()
+
+    def reset_round_counter(self):
+        """Reset the round function counter."""
+        self.n_eval_round = 0
+
+    def _initialize_worker(self, local):
+        """Initialize worker threads."""
+        # create a copy of the objective to maybe be thread-safe.
+        local.objective = deepcopy(self.problem.objective)
```
[the remark this answers is not in the document]
Comment (dweindl) [reply pypesto/optimize/ess.py:192]: > is a deepcopy really needed? could cause overhead.

Depends on what assumptions we want to make on the Objective... 

> amiciobjective should be thread safe, right?

No, you can't use the same solver instance concurrently.

File: pypesto/optimize/ess.py [hunk 7]
```diff
@@ -0,0 +1,662 @@
...
+        self.x[i], self.fx[i] = self.evaluator.single_random()
+        self.n_stuck[i] = 0
+
+
+class ESSOptimizer:
+    """Enhanced Scatter Search (ESS) global optimization.
+
+    .. note: Does not implement any constraint handling yet
+
+    For plausible values of hyperparameters, see VillaverdeEge2012.
```
[the remark this answers is not in the document]
Comment (dweindl) [reply pypesto/optimize/ess.py:300]: One would think so, but sphinx wasn't willing in this very specific case :man_shrugging:

File: pypesto/optimize/ess.py [hunk 8]
```diff
@@ -0,0 +1,662 @@
...
+            x=self.x_best,
+            fval=self.fx_best,
+            message="Global best",
+            **common_result_fields,
+        )
+        # TODO DW: Create a single History with the global best?
+        result.optimize_result.append(optimizer_result)
+
+        # save refset
+        for i in range(self.refset.dim):
```
[the remark this answers is not in the document]
Comment (dweindl) [reply pypesto/optimize/ess.py:452]: Yes. That's up for discussion. Not sure what's best. The overall best for sure, but I am not sure whether one should also include the refset (currently included) and any local minima found along the way (currently not included). Including the refset would make it easier to restart the algorithm, although there is no such option yet...

File: pypesto/optimize/ess.py [hunk 9]
```diff
@@ -0,0 +1,662 @@
...
+        startpoint_method: StartpointMethod,
+        n_threads: int,
+    ):
+        """Construct.
+
+        Parameters
+        ----------
+        problem: The problem
+        startpoint_method: Method for choosing feasible parameters
+        n_threads: Maximum number of threads to use for parallel objective
```
[the remark this answers is not in the document]
Comment (yannikschaelte) [reply pypesto/optimize/ess.py]: full parallelization over cores as opposed to threads? At least for the pypeto.Multi{Process|Thread}Engine we have observed advantages of the former.

File: pypesto/optimize/ess.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (yannikschaelte) [reply pypesto/optimize/ess.py]: at this preliminary point anything including the optimal point is probably good enough. usage may determine what is of interest.

File: pypesto/optimize/ess.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (dweindl) [reply pypesto/optimize/ess.py]: Depends a lot on the Objective. For simple cases, spawning a new process may be more costly than several objective evaluations. For AMICI models, multithreading should work in parallel. For compute-heavy python-only models probably not.

[the remark this answers is not in the document]
Comment (yannikschaelte) [reply pypesto/optimize/ess.py]: ok if it works + gives advantage for amici, the default use case, then this should be good for now. for later, one could also here think about generalizing the parallelization method.

File: pypesto/optimize/ess.py [hunk 10]
```diff
@@ -0,0 +1,662 @@
...
+        """
+        fxs = np.full(shape=n, fill_value=np.nan)
+        xs = np.full(shape=(n, self.problem.dim), fill_value=np.nan)
+
+        while not np.isfinite(fxs).all():
+            retry_indices = np.argwhere(~np.isfinite(fxs)).squeeze()
+            xs[retry_indices] = self.startpoint_method(
+                n_starts=len(retry_indices), problem=self.problem
+            )
+            fxs[retry_indices] = self.multiple(xs)
```
[the remark this answers is not in the document]
Comment (dweindl) [reply pypesto/optimize/ess.py]: :see_no_evil:

**Comment by dweindl** [comment]
@yannikschaelte Comments should be addressed now.

**Comment by paulstapor** [comment]
Nice!
I kinda tried unifying global and local stuff in some toolbox, but didn't mange to continue at some point. Not sure you know it: https://github.com/paulstapor/mary9
Maybe some of the stuff there can help you

**Comment by dweindl** [comment]
> I kinda tried unifying global and local stuff in some toolbox, but didn't mange to continue at some point. Not sure you know it: https://github.com/paulstapor/mary9

Thanks. Did you do any preliminary benchmarking? How were the results?

**Comment by paulstapor** [comment]
> > I kinda tried unifying global and local stuff in some toolbox, but didn't mange to continue at some point. Not sure you know it: https://github.com/paulstapor/mary9
> 
> Thanks. Did you do any preliminary benchmarking? How were the results?

Quite mixed. In some cases, where many failures when evaluating initial points occurred (e.g. when doing pre-equilibration), the preselection actually really helped. In others, only little, sometimes even not at all. I used like 5 rather small models from the Benchmark Collection.

**Comment by paulstapor** [comment]
I'm actually planning to give a talk on my PHD work (with focus on the actual software development part) in my company in roughly one month. I'm trying to find like 1 or 2 motivated people for restarting the work on mary9... No clue if I will find somebody. But if so, the strong advantage ist that all of them are actually very proficient coders... :)
However, progress would remain slowish, since it's just a hobby now, not part of the job as such.

File: test/optimize/test_optimize.py [hunk 11]
```diff
@@ -436,3 +436,36 @@ def test_history_beats_optimizer():
         result_hist.optimize_result.list[0]['fval']
         < result_opt.optimize_result.list[0]['fval']
     )
+
+
+@pytest.mark.parametrize("local_optimizer", [None, optimize.FidesOptimizer()])
+@pytest.mark.flaky(reruns=3)
```
Comment (yannikschaelte) [inline test/optimize/test_optimize.py:442]: bad start :)

**Review by yannikschaelte — APPROVED** [review]

**Comment by yannikschaelte** [comment]
:+1: Looks awesome!

File: test/optimize/test_optimize.py (hunk 11 above)
Comment (dweindl) [reply test/optimize/test_optimize.py:442]: The alternative is checking only that the result is at all better than the starting points as done in the other tests or testing for some very high objective value...

**Pull request merged**